xarray/backends/api.py

```diff
@@ -1827,18 +1842,242 @@ def to_zarr(
+
+    # Needed to handle `store` being a string path
+    store = zarr.hierarchy.normalize_store_arg(
+        store,
+        zarr_version=zarr_version,
+        mode=mode,
+        storage_options=kwargs.get("storage_options", None),
+    )
+    if mode == "w-":
+        # assert that the path does not already exist
+        zarr.open_group(
+            store,
+            mode=mode,
+            storage_options=kwargs.get("storage_options", None),
+            path=kwargs.get("group", None),
+        )
+
+    # Use this to open the group once with all the expected default options
+    # We will reuse xzstore.zarr_group later.
+    xtempstore = backends.ZarrStore.open_group(
+        temp_store,
+        mode="w",  # always write to the temp store
+        zarr_version=zarr_version,
+        consolidated=False,
+        **kwargs,
+    )
+
+    # need to do this separately to get the "coordinates" attribute coorect
+    variables, attrs = conventions.encode_dataset_coordinates(ds)
+    ds = ds._replace(variables=variables, attrs=attrs)
+
+    all_variables = set(ds._variables)
+    # TODO: how do we work with the new index API?
+    index_vars = {dim for dim in ds.dims if dim in all_variables}
+    vars_without_region = {
+        key
+        for key in all_variables - index_vars
+        if (not (set(ds._variables[key].dims) & set(region_dims)))
+    }
+    chunked_vars_without_region = {
+        key for key in vars_without_region if is_chunked_array(ds._variables[key])
+    }
+
+    def extract_encoding(varnames: Iterable[Hashable]) -> dict:
+        return {k: v for k, v in encoding.items() if k in varnames}
+
+    # Always write index variables, and any in-memory variables without region dims
+    eager_write_vars = index_vars | (vars_without_region - chunked_vars_without_region)
+
+    write_to_zarr_store(
+        ds[eager_write_vars],
+        xtempstore,
+        encoding=extract_encoding(eager_write_vars),
+        compute=True,
+        chunkmanager_store_kwargs=kwargs.get("chunkmanager_store_kwargs", None),
+        encode_coordinates=False,
+    )
+
+    # Now initialize the arrays we have not written yet with metadata
+    # but skip any chunked vars without the region, these will get written later
+    vars_to_init = (all_variables - eager_write_vars) - chunked_vars_without_region
+    array_kwargs = {
+        key: kwargs[key]
+        for key in ["safe_chunks", "write_empty", "raise_on_invalid"]
+        if key in kwargs
+    }
+    array_kwargs.setdefault("write_empty", False)
+    if mode == "w":
+        array_kwargs.setdefault("overwrite", True)
+    enc = extract_encoding(vars_to_init)
+    for var in vars_to_init:
+        variable = ds._variables[var]
+        # duplicates dump_to_store
+        if var in enc:
+            variable.encoding = enc[var]
+        encoded = encode_zarr_variable(variable)
+        add_array_to_store(var, encoded, group=xtempstore.zarr_group, **array_kwargs)
```

Comment:
You obv know much more than me here, and I generally use `.drop_encoding` everywhere....

...when I've done similar things, I've just done `ds[vars].to_zarr(mode='a')` for this whole section. Are we doing this ourselves because we need to do more deliberate work on encodings etc?

Reply:
This is just creating the zarr array and not adding any data. These are the "big" variables that get filled in with region writes later. So we end up needing to handle encoding ourselves.

Reply:
(sorry, I had meant `xr.ones_like(ds[vars]).to_zarr(mode='a', compute=False)`). But yes totally makes sense about the encoding, thanks